File: examples/python/gateway_embed.py
from __future__ import annotations

from sipp import (
    CacheRuntimeConfig,
    SippClient,
    ContextRuntimeConfig,
    GatewayDescriptor,
    LocalEmbedOptions,
    LocalModelDescriptor,
    ModelPlacementConfig,
    NativeRuntimeConfig,
    ObservabilityRuntimeConfig,
    ResidencyRuntimeConfig,
    SamplingRuntimeConfig,
    SchedulerRuntimeConfig,
    set_llama_log_quiet,
)

from _support import (
    DEFAULT_CONTEXT,
    DEFAULT_SEED,
    DEFAULT_TEMPERATURE,
    float_env,
    gpu_layers,
    int_env,
    print_embedding,
    read_gateway_args,
    required_env,
)


def runtime_config(*, embeddings: bool) -> NativeRuntimeConfig:
    return NativeRuntimeConfig(
        placement=ModelPlacementConfig(gpu_layers=gpu_layers()),
        context=ContextRuntimeConfig(
            n_ctx=int_env("SIPP_CONTEXT", DEFAULT_CONTEXT),
            n_threads=int_env("SIPP_THREADS"),
            n_threads_batch=int_env("SIPP_THREADS"),
            embeddings=embeddings,
            pooling="mean" if embeddings else None,
        ),
        sampling=SamplingRuntimeConfig(
            temperature=float_env("SIPP_TEMPERATURE", DEFAULT_TEMPERATURE),
            seed=int_env("SIPP_SEED", DEFAULT_SEED),
        ),
        scheduler=SchedulerRuntimeConfig(
            continuous_batching=True,
            prefill_chunk_size=0,
        ),
        cache=CacheRuntimeConfig(mode="live_slot_prefix"),
        residency=ResidencyRuntimeConfig(max_gpu_models_per_device=1),
        observability=ObservabilityRuntimeConfig(runtime_metrics=True),
    )


def main() -> None:
    model, target, input_text = read_gateway_args(
        "gateway_embed", "SippClient gateway embedding example input."
    )
    set_llama_log_quiet(True)

    client = SippClient()
    local_endpoint = client.add(
        "local",
        LocalModelDescriptor(model, runtime_config(embeddings=True)),
    )
    gateway_endpoint = client.add(
        "gateway",
        GatewayDescriptor(
            target,
            required_env("SIPP_GATEWAY_URL"),
            authentication_kind="bearer",
            authentication_value=required_env("SIPP_GATEWAY_TOKEN"),
        )
    )

    local = client.embed(
        input_text,
        endpoint=local_endpoint,
        local=LocalEmbedOptions(
            context_key="python-gateway-embed-local",
            normalize=True,
        ),
    ).result()
    gateway = client.embed(
        input_text,
        endpoint=gateway_endpoint,
    ).result()

    print("local:")
    print_embedding(local)
    print("gateway:")
    print_embedding(gateway)


if __name__ == "__main__":
    main()

File: examples/rust/src/gateway_embed.rs
mod support;

use futures::executor::block_on;
use sipp::backend::set_llama_log_quiet;
use sipp::engine::{
    CacheRuntimeConfig, ContextRuntimeConfig, GpuLayerConfig, KvReuseMode, ModelPlacementConfig,
    NativeRuntimeConfig, ObservabilityRuntimeConfig, PoolingType, ResidencyRuntimeConfig,
    SamplingRuntimeConfig, SchedulerRuntimeConfig,
};
use sipp::{
    GatewayAuthentication, GatewayDescriptor, GatewayRoutes, GatewaySecret, GatewayTimeoutPolicy,
    LocalDescriptor, LocalEmbedOptions, SippClient, SippEmbedRequest,
};

fn main() -> support::ExampleResult<()> {
    block_on(async {
        let args = support::gateway_args(
            "SippClient gateway embedding example input.",
            "gateway_embed",
        )?;
        set_llama_log_quiet(true);

        let mut client = SippClient::new();
        let mut local_descriptor = LocalDescriptor::files([args.model_path]);
        local_descriptor.config = runtime_config(true);
        let local_endpoint = client.add("local", local_descriptor).await?;
        let descriptor = GatewayDescriptor {
            target: args.target.clone(),
            base_url: support::required_env("SIPP_GATEWAY_URL")?,
            routes: GatewayRoutes::default(),
            authentication: GatewayAuthentication::Bearer(GatewaySecret::new(
                support::required_env("SIPP_GATEWAY_TOKEN")?,
            )),
            static_headers: Default::default(),
            timeouts: GatewayTimeoutPolicy::default(),
            protocol_options: Default::default(),
        };
        let gateway_endpoint = client.add("gateway", descriptor).await?;

        let local = client
            .embed(SippEmbedRequest {
                endpoint: Some(local_endpoint),
                input: args.input.clone(),
                local: LocalEmbedOptions {
                    context_key: Some("rust-gateway-embed-local".to_string()),
                    normalize: Some(true),
                },
                ..Default::default()
            })
            .await?;

        let gateway = client
            .embed(SippEmbedRequest {
                endpoint: Some(gateway_endpoint),
                input: args.input,
                ..Default::default()
            })
            .await?;

        println!("local:");
        support::print_embedding(local);
        println!("gateway:");
        support::print_embedding(gateway);
        Ok(())
    })
}

fn runtime_config(embeddings: bool) -> NativeRuntimeConfig {
    NativeRuntimeConfig {
        placement: ModelPlacementConfig {
            gpu_layers: support::env_parse("SIPP_GPU_LAYERS")
                .map(GpuLayerConfig::from_layer_count)
                .unwrap_or(GpuLayerConfig::Auto),
            ..Default::default()
        },
        context: ContextRuntimeConfig {
            n_ctx: support::env_parse("SIPP_CONTEXT").or(Some(support::DEFAULT_CONTEXT)),
            n_threads: support::env_parse("SIPP_THREADS"),
            n_threads_batch: support::env_parse("SIPP_THREADS"),
            embeddings: embeddings.then_some(true),
            pooling: embeddings.then_some(PoolingType::Mean),
            ..Default::default()
        },
        sampling: SamplingRuntimeConfig {
            temperature: support::env_parse("SIPP_TEMPERATURE")
                .or(Some(support::DEFAULT_TEMPERATURE)),
            seed: support::env_parse("SIPP_SEED").or(Some(support::DEFAULT_SEED)),
            ..Default::default()
        },
        scheduler: SchedulerRuntimeConfig {
            continuous_batching: true,
            prefill_chunk_size: 0,
            ..Default::default()
        },
        cache: CacheRuntimeConfig {
            mode: KvReuseMode::LiveSlotPrefix,
            ..Default::default()
        },
        multimodal: Default::default(),
        residency: ResidencyRuntimeConfig {
            max_gpu_models_per_device: 1,
            ..Default::default()
        },
        observability: ObservabilityRuntimeConfig {
            runtime_metrics: true,
            backend_profiling: false,
        },
    }
}
